File: internal/app/aiprovider.go
package app

import (
	"github.com/davidmovas/postulator/internal/domain/entities"
	"github.com/davidmovas/postulator/internal/dto"
	"github.com/davidmovas/postulator/pkg/ctx"
	"github.com/davidmovas/postulator/pkg/errors"
)

func (a *App) CreateAIProvider(req *dto.AIProviderCreate) *dto.Response[string] {
	if req == nil {
		return dtoErr[string](errors.Validation("provider payload is required"))
	}

	entity := &entities.AIProvider{
		Name:     req.Name,
		APIKey:   req.APIKey,
		Provider: req.Provider,
		Model:    req.Model,
		IsActive: req.IsActive,
	}

	if err := a.aiProvSvc.CreateProvider(ctx.FastCtx(), entity); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "created"}
}

func (a *App) GetAIProvider(id int64) *dto.Response[*dto.AIProvider] {
	p, err := a.aiProvSvc.GetProvider(ctx.FastCtx(), id)
	if err != nil {
		return dtoErr[*dto.AIProvider](asAppErr(err))
	}

	return &dto.Response[*dto.AIProvider]{Success: true, Data: dto.FromAIProvider(p)}
}

func (a *App) ListAIProviders() *dto.Response[[]*dto.AIProvider] {
	items, err := a.aiProvSvc.ListProviders(ctx.FastCtx())
	if err != nil {
		return dtoErr[[]*dto.AIProvider](asAppErr(err))
	}

	return &dto.Response[[]*dto.AIProvider]{Success: true, Data: dto.FromAIProviders(items)}
}

func (a *App) ListActiveAIProviders() *dto.Response[[]*dto.AIProvider] {
	items, err := a.aiProvSvc.ListActiveProviders(ctx.FastCtx())
	if err != nil {
		return dtoErr[[]*dto.AIProvider](asAppErr(err))
	}

	return &dto.Response[[]*dto.AIProvider]{Success: true, Data: dto.FromAIProviders(items)}
}

func (a *App) UpdateAIProvider(req *dto.AIProviderUpdate) *dto.Response[string] {
	if req == nil {
		return dtoErr[string](errors.Validation("provider payload is required"))
	}
	entity := &entities.AIProvider{
		ID:       req.ID,
		Name:     req.Name,
		APIKey:   req.APIKey,
		Provider: req.Provider,
		Model:    req.Model,
		IsActive: req.IsActive,
	}

	if err := a.aiProvSvc.UpdateProvider(ctx.FastCtx(), entity); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "updated"}
}

func (a *App) DeleteAIProvider(id int64) *dto.Response[string] {
	if err := a.aiProvSvc.DeleteProvider(ctx.FastCtx(), id); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "deleted"}
}

func (a *App) SetAIProviderStatus(id int64, active bool) *dto.Response[string] {
	if err := a.aiProvSvc.SetProviderStatus(ctx.FastCtx(), id, active); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "updated"}
}

func (a *App) GetAvailableModels(providerName string) *dto.Response[[]string] {
	models := a.aiProvSvc.GetAvailableModels(providerName)
	out := make([]string, 0, len(models))
	for _, m := range models {
		out = append(out, string(m))
	}

	return &dto.Response[[]string]{Success: true, Data: out}
}

func (a *App) ValidateModel(providerName, model string) *dto.Response[string] {
	if err := a.aiProvSvc.ValidateModel(providerName, model); err != nil {
		return dtoErr[string](asAppErr(err))
	}

	return &dto.Response[string]{Success: true, Data: "valid"}
}

File: frontend/src/services/aiProvider.ts
import {
  CreateAIProvider,
  DeleteAIProvider,
  GetAIModels,
  GetAIProvider,
  GetAvailableModels,
  ListAIProviders,
  ListActiveAIProviders,
  SetAIProviderStatus,
  UpdateAIProvider,
  ValidateModel,
} from "@/wailsjs/wailsjs/go/app/App";
import { dto } from "@/wailsjs/wailsjs/go/models";
import { unwrapMany, unwrapOne, unwrapString } from "./utils";

export interface AIProvider {
  id: number;
  name: string;
  provider: string;
  model: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface ModelsByProvider {
  openai: string[];
  anthropic: string[];
  google: string[];
}

function mapAIProvider(x: dto.AIProvider): AIProvider {
  return {
      id: x.id,
      name: x.name,
      provider: x.provider,
      model: x.model,
      isActive: x.isActive,
      createdAt: x.createdAt,
      updatedAt: x.updatedAt,
  };
}

function mapModelsByProvider(x: dto.ModelsByProvider): ModelsByProvider {
  return { openai: x.openai, anthropic: x.anthropic, google: x.google };
}

export async function listAIProviders(): Promise<AIProvider[]> {
  const res = await ListAIProviders();
  return unwrapMany<dto.AIProvider>(res).map(mapAIProvider);
}

export async function listActiveAIProviders(): Promise<AIProvider[]> {
  const res = await ListActiveAIProviders();
  return unwrapMany<dto.AIProvider>(res).map(mapAIProvider);
}

export async function getAIProvider(id: number): Promise<AIProvider> {
  const res = await GetAIProvider(id);
  return mapAIProvider(unwrapOne<dto.AIProvider>(res));
}

export async function createAIProvider(input: { name: string; apiKey: string; provider: string, model: string; isActive: boolean }): Promise<string> {
  const payload = new dto.AIProviderCreate({
      name: input.name,
      apiKey: input.apiKey,
      provider: input.provider,
      model: input.model,
      isActive: input.isActive,
  });
  const res = await CreateAIProvider(payload);
  return unwrapString(res);
}

export async function updateAIProvider(input: { id: number; name: string; apiKey?: string;  provider: string, model: string; isActive: boolean }): Promise<string> {
  const payload = new dto.AIProviderUpdate({
      id: input.id,
      name: input.name,
      apiKey: input.apiKey,
      provider: input.provider,
      model: input.model,
      isActive: input.isActive,
  });
  const res = await UpdateAIProvider(payload);
  return unwrapString(res);
}

export async function deleteAIProvider(id: number): Promise<string> {
  const res = await DeleteAIProvider(id);
  return unwrapString(res);
}

export async function getAIModels(): Promise<ModelsByProvider> {
  const res = await GetAIModels();
  return mapModelsByProvider(unwrapOne<dto.ModelsByProvider>(res as any));
}

export async function getAvailableModels(providerName: string): Promise<string[]> {
  const res = await GetAvailableModels(providerName);
  return unwrapMany<string>(res as any);
}

export async function validateModel(providerName: string, model: string): Promise<string> {
  const res = await ValidateModel(providerName, model);
  return unwrapString(res);
}

export async function setAIProviderStatus(id: number, isActive: boolean): Promise<string> {
  const res = await SetAIProviderStatus(id, isActive);
  return unwrapString(res);
}
